Update the sdk target to net10.0

v-wuzhai opened the pull request.

Commit:
Update the sdk target to net10.0

Commit:
Merge branch 'main' of https://github.com/dotnet/sdk into dev/Jason/Update-the-sdk-target-to-net10.0

ViktorHofer commented on Directory.Build.props:
```diff
@@ -34,7 +34,7 @@
     <IncludeSymbols>true</IncludeSymbols>
     <IsShippingPackage>false</IsShippingPackage>
-    <SdkTargetFramework>net9.0</SdkTargetFramework>
+    <SdkTargetFramework>net10.0</SdkTargetFramework>
```
I think you also want to set `NetCurrent` here to `net10.0` as at least one project (dotnet-format) uses that property.

mthalman commented on src/SourceBuild/content/Directory.Build.props:
```diff
@@ -110,7 +110,7 @@
 
     <!-- TargetFrameworkDefaults.props -->
-    <NetCurrent>net9.0</NetCurrent>
+    <NetCurrent>net10.0</NetCurrent>
```
We can't make this change yet. This will cause source build to build its bootstrapping tools targeting net10.0, using a 9.0 SDK (via global.json). So it's getting this error in the prep script:

`/vmr/.dotnet/sdk/9.0.100-preview.7.24407.12/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(166,5): error NETSDK1045: The current .NET SDK does not support targeting .NET 10.0.  Either target .NET 9.0 or lower, or use a version of the .NET SDK that supports .NET 10.0. Download the .NET SDK from https://aka.ms/dotnet/download [/vmr/eng/tools/BinaryToolKit/BinaryToolKit.csproj]`

This change couldn't be made until the global.json is updated to target an alpha version of 10.0.

Commit:
Merge branch 'main' of https://github.com/dotnet/sdk into dev/Jason/Update-the-sdk-target-to-net10.0

Commit:
Unlock the build

v-wuzhai commented:
There are a bunch of template test failure error messages as follows: 

> **warning MSB4011**: "/Users/runner/work/1/s/artifacts/.nuget/packages/microsoft.dotnet.arcade.sdk/10.0.0-beta.24504.4/tools/Tests.targets" cannot be imported again. It was already imported at "/Users/runner/work/1/s/artifacts/.nuget/packages/microsoft.dotnet.arcade.sdk/10.0.0-beta.24504.4/tools/Imports.targets (12,3)". This is most likely a build authoring error. This subsequent import will be ignored. [/Users/runner/work/1/s/artifacts/tmp/Release/dotnet-new.IntegrationTests/Restore_Basic/20241008062742474/Target/Output/MyProject.csproj]\n/Users/runner/work/1/s/artifacts/.nuget/packages/microsoft.dotnet.arcade.sdk/10.0.0-beta.24504.4/tools/GenerateChecksums.targets(4,71):
**error MSB4022:** The result "" of evaluating the value "$(ArcadeSdkBuildTasksAssembly)" of the "AssemblyFile" attribute in element <UsingTask> is not valid. [/Users/runner/work/1/s/artifacts/tmp/Release/dotnet-new.IntegrationTests/Restore_Basic/20241008062742474/Target/Output/MyProject.csproj]\nStdErr:\nRestore failed.\nPost action failed.\nManual instructions: Run 'dotnet restore'\n

@dotnet/templating-engine-maintainers @dotnet/arcade-contrib Could you please take a look at this issue?

v-wuzhai commented on Directory.Build.props:
```diff
@@ -34,12 +34,12 @@
 
     <!-- We used to have scenarios where the MSBuild host (VSMac) had an older .NET, but don't any more. -->
-    <ResolverTargetFramework>$(SdkTargetFramework)</ResolverTargetFramework>
+    <ResolverTargetFramework>net9.0</ResolverTargetFramework>
```
Here is the build error message when set to net10.0.

`C:\Users\AzureTest\Desktop\109\sdk\.dotnet\sdk\10.0.100-alpha.1.24507.21\Microsoft.Common.CurrentVersion.targets(1889,5): error NU1702: ProjectReference 'C:\Users\AzureTest\Desktop\109\sdk\test\Microsoft.NET.TestFramework\Microsoft.NET.TestFramework.csproj' was resolved using '.NETFramework,Version=v4.7.2' instead of the project target framework '.NETCoreApp,Version=v9.0'. This project may not be fully compatible with your project.
    C:\Users\AzureTest\Desktop\109\sdk\.dotnet\sdk\10.0.100-alpha.1.24507.21\Microsoft.Common.CurrentVersion.targets(1889,5): error NU1702: ProjectReference 'C:\Users\AzureTest\Desktop\109\sdk\src\Cli\Microsoft.DotNet.Cli.Utils\Microsoft.DotNet.Cli.Utils.csproj' was resolved using '.NETFramework,Version=v4.7.2' instead of the project target framework '.NETCoreApp,Version=v9.0'. This project may not be fully compatible with your project.`

marcpopMSFT replied:
What build did that trigger in? I tried locally and targeting net10.0 worked. The current builds are failing because the test framework isn't building because it doesn't have a net10.0 target (but projects that depend on it do have that). That means the nuget.config for the test projects never gets created. I pushed a change to switch back to net10.0 so we can see where this failure comes from.

mmitche commented:
@kasperk81 I think the Arcade issue is a red herring. When the tests are being run, the .NET SDK imports the Directory.Build.targets from the root of the SDK repo, which then imports the arcade targets. I am pretty sure this is not intended for template tests as they aren't really testing the simple create+restore flow. In a helix run, I don't think this ends up happening because the SDK's repo content isn't on the machine, so that's good.

There is a `Directory.Build.props` file that is set up as part of the test run in the test temp dir. I think there needs to be a corresponding `Directory.Build.targets` alongside it to avoid accidentally importing Directory.Build..targets

/cc @nagilson and @marcpopMSFT

mmitche committed:
Add an empty Directory.Build.targets

mmitche committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

mmitche commented:
@kasperk81 It's due to this change:

https://github.com/dotnet/sdk/pull/43015/files#diff-37aabd56f54205e41e1a83387bfefad4ff62b60b0ec360e3beb1160d1c635722R20

That causes the targets in src/Tests/Microsoft.NET.TestFramework/SetupTestRoot.targets to not run. I think the conditionals in that file look pretty suspect. I'm guessing they were working around some race condition.

marcpopMSFT committed:
Switch the resolvers and test project to net10.0

mmitche commented:
dotnet-format uses NetCurrent right now. I think it should be using SdkTargetFramework

mmitche committed:
Move dotnet-format to use SdkTargetFramework

marcpopMSFT commented:
Looks like we'll have to update some dotnet new baselines as the web templates have already updated to support net10 (though a few things look off there). Additionally, we need to update the global.json for source build. I have the changes locally but I figured I'd wait a bit and see if we get any results from the sdk tests

marcpopMSFT committed:
Fix the template tests but we need to check on this as the current output is wrong as it's missing net9.

marcpopMSFT committed:
Update source build global.json

marcpopMSFT committed:
Update the current target to 10.0 as some tests automatically update to that when the default TFM changes

marcpopMSFT committed:
Fix the formatting of some of the templating tests

marcpopMSFT committed:
Web templates have moved to target net10.0 already

marcpopMSFT committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

Commit:
Update according to the comments

Commit:
Fix version

Commit:
The current .NET SDK does not support targeting .NET 11.0

Commit:
Attempting to unlock a bunch of test failures

marcpopMSFT committed:
Temporarily disable a test as we can't reference across TFMs. This can be reenabled once the test tools templates target net10.0

marcpopMSFT commented:
It's possible it may have been better to retarget the templates to net10 before targeting the SDK at net10 per some of the failures. I think that just means more tests we'll have to disable temporarily until the templates support it.

marcpopMSFT committed:
Update the regex to handle 5+

marcpopMSFT committed:
After talking to baronfel, disabling some container tests for now until the templates are all updated.

marcpopMSFT commented:
@dotnet/illink Macos failure in `Microsoft.NET.Publish.Tests.GivenThatWeWantToPublishAnAotApp.NativeAot_hw_runs_with_no_warnings_when_PublishAot_is_enabled(targetFramework: \"net10.0\")`
```
The command output contained a result it should not have contained: warning
...
Generating native code
ld: warning: __LD,__compact_unwind entries for _RhpNewFast overlap at offset 0x32
ld: warning: __LD,__compact_unwind entries for _RhpNewObject overlap at offset 0x6B
...
```

A bunch of warnings when generating the native code which the test flags.

marcpopMSFT commented:
@dotnet/product-construction `error NU1301: Unable to load the service index for source https://pkgs.dev.azure.com/ms/BuildXL/_packaging/BuildXL/nuget/v3/index.json.` in the VMR legs.

am11 commented:
```
The command output contained a result it should not have contained: warning

File Name: /tmp/helix/working/BB8C09A0/p/d/dotnet
Arguments: msbuild /t:Publish /private/tmp/helix/working/BB8C09A0/w/A1EB088D/e/testExecutionDirectory/NativeAot_hw_---788461B1_1/HellowWorldNativeAotApp/HellowWorldNativeAotApp.csproj /restore /p:UseCurrentRuntimeIdentifier=true /p:SelfContained=true /p:CheckEolTargetFramework=false
Exit Code: 0
StdOut:
MSBuild version 17.13.0-preview-24504-04+c4d51a11b for .NET
Determining projects to restore...
Restored /private/tmp/helix/working/BB8C09A0/w/A1EB088D/e/testExecutionDirectory/NativeAot_hw_---788461B1_1/HellowWorldNativeAotApp/HellowWorldNativeAotApp.csproj (in 382 ms).
/tmp/helix/working/BB8C09A0/p/d/sdk/10.0.100-ci/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(326,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/private/tmp/helix/working/BB8C09A0/w/A1EB088D/e/testExecutionDirectory/NativeAot_hw_---788461B1_1/HellowWorldNativeAotApp/HellowWorldNativeAotApp.csproj]
HellowWorldNativeAotApp -> /private/tmp/helix/working/BB8C09A0/w/A1EB088D/e/testExecutionDirectory/NativeAot_hw_---788461B1_1/HellowWorldNativeAotApp/bin/Debug/net10.0/osx-x64/HellowWorldNativeAotApp.dll
Generating native code
ld: warning: __LD,__compact_unwind entries for _RhpNewFast overlap at offset 0x32
ld: warning: __LD,__compact_unwind entries for _RhpNewObject overlap at offset 0x6B
ld: warning: __LD,__compact_unwind entries for _RhNewString overlap at offset 0x61
ld: warning: __LD,__compact_unwind entries for _RhpNewArray overlap at offset 0x6C
ld: warning: __LD,__compact_unwind entries for _RhpNewArrayRare overlap at offset 0x6F
ld: warning: __LD,__compact_unwind entries for _RhpThrowHwEx overlap at offset 0x6B
ld: warning: __LD,__compact_unwind entries for _RhpThrowEx overlap at offset 0x9F
ld: warning: __LD,__compact_unwind entries for _RhpRethrow overlap at offset 0x65
ld: warning: __LD,__compact_unwind entries for _RhpCallCatchFunclet overlap at offset 0xC0
ld: warning: __LD,__compact_unwind entries for _RhpCallFinallyFunclet overlap at offset 0xAB
ld: warning: __LD,__compact_unwind entries for _RhpCallFilterFunclet overlap at offset 0x32
ld: warning: __LD,__compact_unwind entries for _RhpCallPropagateExceptionCallback overlap at offset 0xCA
ld: warning: __LD,__compact_unwind entries for _RhpWaitForGC overlap at offset 0x96
ld: warning: __LD,__compact_unwind entries for _RhpGcPollRare overlap at offset 0x43
ld: warning: __LD,__compact_unwind entries for _RhpStackProbe overlap at offset 0x1F
ld: warning: __LD,__compact_unwind entries for _RhpPInvoke overlap at offset 0x33
ld: warning: __LD,__compact_unwind entries for _RhpUniversalTransition overlap at offset 0xDE
ld: warning: __LD,__compact_unwind entries for RhpUniversalTransition_DebugStepTailCall overlap at offset 0xDE
HellowWorldNativeAotApp -> /private/tmp/helix/working/BB8C09A0/w/A1EB088D/e/testExecutionDirectory/NativeAot_hw---788461B1_1/HellowWorldNativeAotApp/bin/Debug/net10.0/osx-x64/publish/
```

@filipnavara these warnings are showing up on osx-x64 13. Do they look familiar?

filipnavara commented:
> @filipnavara these warnings are showing up on osx-x64 13. Do they look familiar?

Unfortunately no, they don't. It's the native assembly code and it's coming from this macro: https://github.com/dotnet/runtime/blob/6d23f64952d0f242c4dcc7cbb230a619be930309/src/coreclr/nativeaot/Runtime/unix/unixasmmacrosamd64.inc#L19-L27. Weirdly we don't emit the same block on arm64, where it would equally make sense to ensure that LLVM or LD64 doesn't try to convert the DWARF unwinding codes into the compact unwinding code (which is buggy for non-standard unwind sequences).

I haven't done osx-x64 builds for a while. Do we know what Xcode version is on the machine? (should be the clang version in the log I guess)

am11 commented:
> Do we know what Xcode version is on the machine? (should be the clang version in the log I guess)

I think the reason why it is failing in this PR and not main branch is due to some changes between 9.0 rc2 and 10.0 alpha1. Which means both runtime and SDK repos are now on LKG testing plan, so even the unintentional breaking changes can make it to SDK release unchecked. The integration tests will only spot the issues when global.json / SDK is updated in either runtime or SDK repos. cc @dotnet/ilc-contrib

marcpopMSFT commented:
@dotnet/sdk-container-builds-maintainers there are still a few container tests failing. Most are from the rate limiting so rerunning would probably be sufficient. Not sure about `Microsoft.NET.Build.Containers.IntegrationTests.EndToEndTests.EndToEnd_NoAPI_ProjectType(projectType: \"webapi\", addPackageReference: True)
Error: No such container: c0c2cdf623077d9341b7ed228db6f2cadef3bbf216370e935be732ba47129e2e`

marcpopMSFT committed:
Disable a couple more container tests

am11 committed:
Add ignoredPatterns in NotHaveStdOutContaining

am11 committed:
Ignore __compact_unwind warnings

am11 commented:
We can likely disregard these `__unwind` warnings: https://github.com/dotnet/sdk/compare/main...am11:sdk:ignore-ld-warning, unless someone is particularly interested in investigating it further and going down the rabbit hole.

Justification:
1. The `osx-x64` platform is gradually becoming less relevant: https://github.com/dotnet/runtime/issues/108570
2. The older version of `clang` on `osx-x64` is even less of a priority.
3. These warnings (as far as we can tell) don't have any observable impact on functionality, exception handling or debugging.

@filipnavara, @MichalStrehovsky what do you guys think?

am11 committed:
Ignore clang-16 warning as well

jkotas commented:
> This PR and not main branch is due to some changes between 9.0 rc2 and 10.0 alpha1.

What is the change between 9.0 rc2 and 10.0 alpha1 that introduced this? It would be useful to connect the dots. The best fix would be to revert the offending change if possible.

> Which means both runtime and SDK repos are now on LKG testing plan

I am not sure what you mean. The LKG changes in dotnet/runtime only apply to the toolset used for building the compilers. The tests are all live bits. I think we have not seen this in dotnet/runtime due to different XCode versions used for testing or not checking for XCode warnings when running the tests.

am11 commented:
> I think we haven’t seen this in `dotnet/runtime` due to different Xcode versions used for testing or not checking for Xcode warnings when running tests.

Yup, since several `dotnet/runtime` to `dotnet/sdk` codeflow PRs were merged between rc2 and this PR, it shows the failing test isn’t using ILCompiler from packages being updated, but rather the version bundled with SDK specified in `global.json`. As you can see that this PR is only updating `global.json` related to the failing test. We skipped over RTM/GA (this PR jumps from rc2 to alpha1), so these tests haven’t run with RTM. I think this is unintentional and a concerning gap.

> What’s the change between 9.0 rc2 and 10.0 alpha1?

Locally, with Xcode 16, I only get one warning (which will also fail this sensitive test—test expects zero warnings on stdout and stderr):

```bash
ld: warning: -ld_classic is deprecated and will be removed in a future release
```

I’m downloading Xcode 13 to see if I can reproduce this warning locally. 

Since toolchain versions aren’t printed in Helix runs, I think `runtime` uses AppleClang 14, and `SDK` uses v13 on osx.13. We can align helix queues between runtime and SDK: `OSX.1200.Amd64.Open` for PRs and `OSX.1200.Amd64` for official builds: 
https://github.com/dotnet/sdk/blob/1c88a979b7109d83e5317b0e374c8d13ef2b3b68/.vsts-pr.yml#L65

jkotas commented:
> the failing test isn’t using ILCompiler from packages being updated, but rather the version bundled with SDK specified in global.json. 

I don't think that's correct. The tests are using ILCompiler package 10.0.0-alpha.1.24507.22. This version is specified in https://github.com/dotnet/sdk/blob/main/eng/Versions.props.

I think that we are seeing this failure in this PR since this PR is enabling the tests for .NET 10 TFM.

You can run `runfo get-helix-payload -j  f20fa4a6-963c-4538-80f5-dd163343cca5 -w 1ed3f53c-bf12-4be1-8b68-1ab87b11983b -o c:\helix_payload` to get the exact bits that are downloaded to the test machine to tell what's going on.

am11 commented:
> The tests are using ILCompiler package 10.0.0-alpha.1.24507.22. This version is specified in [main/eng/Versions.props](https://github.com/dotnet/sdk/blob/main/eng/Versions.props?rgh-link-date=2024-10-21T15%3A28%3A40Z).

This failing test https://github.com/dotnet/sdk/blob/1c88a979b7109d83e5317b0e374c8d13ef2b3b68/test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs#L24-L78 is using SDK bundled version. I think the example you are referring to would look like https://github.com/dotnet/sdk/blob/1c88a979b7109d83e5317b0e374c8d13ef2b3b68/test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs#L267 but the failing test is not using PackageReference.

jkotas commented:
Yes, the test is using the `10.0.0-alpha.1.24507.22` ILCompiler version embedded into the `10.0.100-ci` live-built SDK.

am11 commented:
It has:
```
     GetKnownILCompilerPackVersion(testAsset, targetFramework, out string expectedVersion); 
     CheckIlcVersions(testAsset, targetFramework, rid, expectedVersion); 
```
it is explicitly checking for known ILCompiler pack which my understanding is the bundled version.

Passing on main, failing in PR that is modifying global.json. Other tests in the same file are not affected.

jkotas committed:
Add temp logging

jkotas committed:
Add .NET 9.0 TFM to PublishTestUtils

jkotas commented:
@jkoritzinsky We have picked up more failures in this PR that seems to be caused by https://github.com/dotnet/runtime/pull/107889 . Could you please take a look?

From https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-sdk-refs-pull-43015-merge-5284031d0a3746c894/Microsoft.NET.Sdk.Web.Tests.dll/3/console.03fc7d48.log?skoid=8eda00af-b5ec-4be9-b69b-0919a2338892&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2024-10-25T02%3A22%3A13Z&ske=2024-10-25T03%3A22%3A13Z&sks=b&skv=2024-05-04&sv=2024-05-04&st=2024-10-25T02%3A22%3A13Z&se=2024-10-25T03%3A22%3A13Z&sr=c&sp=r&sig=hngjvu62hltBJRfd%2Fk9BtGpwru3oo8QaCEyGaKZIyc0%3D:

```
[m[37m        Generating native code
[m[37m        /usr/bin/ld.bfd: /usr/bin/ld.bfd: DWARF error: invalid or unhandled FORM value: 0x23
[m[37m        /datadisks/disk1/work/997E08B4/w/A9900972/e/testExecutionDirectory/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/10.0.0-alpha.1.24523.5/sdk/libeventpipe-enabled.a(unity_0_cxx.cxx.o): in function `ds_dump_protocol_helper_handle_ipc_message(_DiagnosticsIpcMessage*, _DiagnosticsIpcStream*)':
[m[37m        unity_0_cxx.cxx:(.text._Z42ds_dump_protocol_helper_handle_ipc_messageP22_DiagnosticsIpcMessageP21_DiagnosticsIpcStream+0x2ed): undefined reference to `minipal_get_length_utf16_to_utf8'
[m[37m        /usr/bin/ld.bfd: unity_0_cxx.cxx:(.text._Z42ds_dump_protocol_helper_handle_ipc_messageP22_DiagnosticsIpcMessageP21_DiagnosticsIpcStream+0x325): undefined reference to `minipal_convert_utf16_to_utf8'
[m[37m        /usr/bin/ld.bfd: unity_0_cxx.cxx:(.text._Z42ds_dump_protocol_helper_handle_ipc_messageP22_DiagnosticsIpcMessageP21_DiagnosticsIpcStream+0x47e): undefined reference to `minipal_get_length_utf8_to_utf16'
[m[37m        /usr/bin/ld.bfd: unity_0_cxx.cxx:(.text._Z42ds_dump_protocol_helper_handle_ipc_messageP22_DiagnosticsIpcMessageP21_DiagnosticsIpcStream+0x4c5): undefined reference to `minipal_convert_utf8_to_utf16'
...
```

jkoritzinsky commented:
Yes, I'll take a look.

jkoritzinsky commented:
Looks like we now have a circular dependency between the runtime lib and the eventpipe lib by putting the minipal in the runtime lib. I think the easiest solution here (other than adding the flags to do a second linker pass, which I think we want to avoid), is to ship the minipal as a separate static lib in the package for NativeAOT. I'll work on getting that working.

jkotas commented:
> it is explicitly checking for known ILCompiler pack which my understanding is the bundled version.

It is the bundled version in the live built of the SDK. It is independent on the version in global.json.

jkotas commented:
> What is the change between 9.0 rc2 and 10.0 alpha1 that introduced this?

This is what I found: These warnings show up when dotnet/runtime compiled using Apple clang 15+ is used with classic linker (either Apple clang 14 or clang 15+ with -ld_classic) for PublishAot. The warnings got introduced as side-effect of https://github.com/dotnet/runtime/pull/107871 that updated Apple clang version used to build dotnet/runtime main branch from 14 to 15.

jkotas committed:
Revert "Add temp logging"

This reverts commit 252429dc1be46d054908ef7810ee3ad99c8a52fb.

am11 commented:
> It is the bundled version in the live build of the SDK. It is independent on the version in global.json.

Right, https://github.com/dotnet/runtime/pull/107871 and https://github.com/dotnet/runtime/pull/107889 are already part of the daily build releases: https://github.com/dotnet/sdk/blob/main/documentation/package-table.md (10.0.x) and they did not fail coherency updates PR. We can already build projects with `<TargetFramework>net10.0`. If these were bugs, they weren't caught during the coherency update.

jkotas commented:
> If these were bugs, they weren't caught during the coherency update.

Right, that's expected given how things are setup in this repo. There are no publish tests running against .NET 10 TFM in dotnet/sdk main. Ingesting .NET 10 runtime packages into sdk have not automatically enabled all tests to run against .NET 10 TFM. Changes in this PR (`ToolsetInfo.CurrentTargetFramework`) are enabling the publish tests to run against .NET 10 TFM and that's why we are seeing the failures.

jkotas committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

jkotas committed:
Avoid test directory name collisions

Tests with targetFramework theories have to use unique directory for
each targetFramework. The Arcade retry logic happened to compensate for
up to 2 colliding directories, but that does not work anymore now that
we have one more targetFramework.

jkotas committed:
Merge remote-tracking branch 'am11/ignore-ld-warning' into dev/Jason/Update-the-sdk-target-to-net10.0

jkotas committed:
Add comments

jkotas commented:
> We can likely disregard these __unwind warnings: https://github.com/dotnet/sdk/compare/main...am11:sdk:ignore-ld-warning

@am11 Thank you for proposing a fix. I have merged your change into this PR and added some explanatory comments.

ViktorHofer commented on src/SourceBuild/content/global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "9.0.100-rc.1.24452.12"
+    "dotnet": "10.0.100-alpha.1.24507.21"
```
Undo this change. We can't build with a 10.0.100 SDK yet in the VMR.

am11 replied:
We can try applying https://github.com/dotnet/runtime/pull/109265.patch to see if there anything else blocking the flow.

marcpopMSFT replied:
Not sure we can have the SDK target 10 and not use the net10 sdk? Do we need to disable the VMR until the support for the 10 sdk is enabled?

ViktorHofer replied:
True, if a repo that is part of the VMR builds with a newer sdk and targets a newer TFM, then the VMR's SDK needs to be updated as well.

I will look into this tomorrow. We might just need to set a few additional temporary global properties when building runtime.

ViktorHofer committed:
Set NetCurrent accordingly and downgrade VMR SDK

ViktorHofer committed:
Set NetCoreAppToolCurrent to SDK value

ViktorHofer committed:
Forward NetCoreAppToolCurrentVersion

ViktorHofer committed:
Update 0001-Forward-NetCoreAppToolCurrent-switch.patch

ViktorHofer committed:
Fix patch

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

am11 commented:
@ViktorHofer. (re-pitching the idea 🙂) I think if we continue to apply the RollForward patches (which @jkoritzinsky already did for cdac project https://github.com/dotnet/runtime/pull/108946 and it evidently fixed that part of the build), I think runtime build will clear. When everything is on net10.0 plan, we can delete those patches.

ps - Ideally https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs shouldn't be invoking `dotnet crossgen2.dll` in the first place; instead it should be using the published `crossgen2.exe`, which we consume on end-user's machine from the nuget package. That crossgen2.exe from the nuget package is either PublishSingleFile or PublishAot and there is no crossgen2.dll variant in realworld/shipped nupkgs.

ViktorHofer committed:
Add patch for hardcoded LinkTask TFM

ViktorHofer committed:
update patches because of sync failure

ViktorHofer committed:
CLRF -> LF in patch

jkotas commented:
> instead it should be using the published crossgen2.exe, which we consume on end-user's machine from the nuget package. That crossgen2.exe from the nuget package is either published with PublishSingleFile or PublishAot (depending on the platform's AOT support) and there is no crossgen2.dll variant in realworld/shipped nupkgs.

This would mean that we need to do build the live crossgen2 package in the middle of coreclr build and it would be a throw-away package for cross-compilation. This approach was explored a bit in https://github.com/dotnet/runtime/pull/107772, but rejected.

am11 commented:
> This would mean that we need to do build the live crossgen2 package in the middle of coreclr build and it would be a throw-away package for cross-compilation.

LKG has published crossgen2.exe and the crossgen2.csproj which gets published right after the coreclr build (before libs subset) also has the same published crossgen2.exe. From that perspective we are covered. I think when RunReadyToRunCompiler task is executed, we have the published crossgen2.exe available either from LKG or the just built version, no need for additional step. (side note: there is an unused copy of RunReadyToRunCompiler task in runtime repo which is outdated, only the one in SDK repo is maintained and used)

ViktorHofer commented:
```
    D:\a\_work\1\vmr\src\runtime\src\libraries\System.Text.Json\src\System\Text\Json\JsonHelpers.cs(226,28): error CS1061: 'Dictionary<string, TValue>' does not contain a definition for 'GetAlternateLookup' and no accessible extension method 'GetAlternateLookup' accepting a first argument of type 'Dictionary<string, TValue>' could be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\vmr\src\runtime\src\libraries\System.Text.Json\src\System.Text.Json.csproj::TargetFramework=net9.0]
    D:\a\_work\1\vmr\src\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Serialization\Converters\Value\EnumConverter.cs(307,112): error CS1061: 'Dictionary<string, EnumConverter<T>.EnumFieldInfo>' does not contain a definition for 'GetAlternateLookup' and no accessible extension method 'GetAlternateLookup' accepting a first argument of type 'Dictionary<string, EnumConverter<T>.EnumFieldInfo>' could be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\vmr\src\runtime\src\libraries\System.Text.Json\src\System.Text.Json.csproj::TargetFramework=net9.0]
```

cc @stephentoub @jkotas this happens when building runtime with a .NET 10 SDK.

jkotas commented:
> this happens when building runtime with a .NET 10 SDK.

This means that the .NET 9 reference assemblies used by the build are stale. What's the .NET 9 ref-pack version used by the VMR build?

ViktorHofer commented:
/vmr/src/runtime/artifacts/sb/package-cache/microsoft.netcore.app.ref/**9.0.0-preview.7.24405.7**/ref/net9.0/System.Text.Encoding.CodePages.dll

am11 commented:
https://github.com/dotnet/sdk/blob/3b653cc71df279eeb0050e427504b3d4fc40f60e/src/Installer/redist-installer/targets/GenerateBundledVersions.targets#L88-L90

Sounds like it's time to address this TODO :)

ViktorHofer committed:
Update hardcoded tareting and runtime pack versions

ViktorHofer commented:
Yeah found it and just updated all targeting and runtime packs to RC2.

ViktorHofer commented:
This change won't reach the VMR repos though as they build with an LKG SDK. We need to update them in a separate PR.

ViktorHofer committed:
TestCLI package is net9.0

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

ViktorHofer commented:
We are now waiting for this build to finish: https://dnceng.visualstudio.com/internal/_build/results?buildId=2571337&view=results

... and afterwards update the two global.json files in this PR to the new SDK version.

I'm logging off now. If someone could do that in ~1 hour, that would be fantastic.

am11 commented on src/SourceBuild/content/global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "9.0.100-rc.1.24452.12"
+    "dotnet": "10.0.100-alpha.1.24507.21"
```
```suggestion
    "dotnet": "10.0.100-alpha.1.24529.16"
```

am11 commented on global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "9.0.100-rc.2.24474.11",
+    "dotnet": "10.0.100-alpha.1.24507.21",
```
```suggestion
    "dotnet": "10.0.100-alpha.1.24529.16",
```

am11 commented:
I have double checked that this version includes @ViktorHofer's commit:

```zsh
% curl -sSL https://aka.ms/dotnet/10.0.1xx/daily/productCommit-win-x64.json | jq -r '.sdk'  
{
  "commit": "ded76b9f2e3cb8bc29509f022e9bf6d2abe2e3eb",
  "version": "10.0.100-alpha.1.24529.16"
}
```

https://github.com/dotnet/sdk/commits/ded76b9f2e3cb8bc29509f022e9bf6d2abe2e3eb

jkotas committed:
Apply suggestions from code review

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

marcpopMSFT commented:
VMR build still failing:
```
  /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS1620: Argument 2 must be passed with the 'out' keyword [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
  /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS0269: Use of unassigned out parameter 'task' [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
  /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,137): error CS0177: The out parameter 'task' must be assigned to before control leaves the current method [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
  

```

jkoritzinsky commented:
@elinor-fung looks like there's some VMR failures in the cdac work when building against .NET 10.

ViktorHofer commented:
@dotnet/source-build any idea why the RC2 targeting and runtime pack is missing from these feeds? Do we need to update 

```
/vmr/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Ref with version (= 9.0.0-rc.2.24473.5) [/vmr/build.proj]
/vmr/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj : error NU1102:   - Found 7 version(s) in /vmr/prereqs/packages/reference/ [ Nearest version: 8.0.0 ] [/vmr/build.proj]
/vmr/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj : error NU1102:   - Found 1 version(s) in /vmr/prereqs/packages/prebuilt/ [ Nearest version: 9.0.0-rc.1.24431.7 ] [/vmr/build.proj]
/vmr/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection/Microsoft.DotNet.SourceBuild.Tasks.LeakDetection.csproj : error NU1102:   - Found 1 version(s) in /vmr/prereqs/packages/previously-source-built/ [ Nearest version: 9.0.0-preview.7.24405.7 ] [/vmr/build.proj]
```

I see that the VMR is still using an RC1 SDK here: https://github.com/dotnet/sdk/blob/61fed1eeffdd37fe9336273f39fe7c2c6cb43cfa/src/SourceBuild/content/global.json#L3
And most of the SourceBuilt packages are on P7: https://github.com/dotnet/sdk/blob/61fed1eeffdd37fe9336273f39fe7c2c6cb43cfa/src/SourceBuild/content/eng/Versions.props#L26-L28

Why didn't they get updated to RC2?

MichaelSimons commented:
> Why didn't they get updated to RC2? 

This is flowing in now with https://github.com/dotnet/sdk/pull/44212.

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

ViktorHofer commented:
While we wait for @jkoritzinsky's fix to flow, let's look at the source-build failures:

```
    /vmr/src/runtime/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/QCallHandles.cs(39,26): warning CS9265: Field 'ByteRef._ref' is never ref-assigned to, and will always have its default value (null reference) [/vmr/src/runtime/src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj]
    /vmr/src/runtime/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs(1168,30): warning CS8619: Nullability of reference types in value of type 'string[]' doesn't match target type 'Span<string?>'. [/vmr/src/runtime/src/libraries/System.Net.Http/src/System.Net.Http.csproj::TargetFramework=net10.0-linux]
    /vmr/src/runtime/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs(1208,42): warning CS8620: Argument of type 'string[]' cannot be used for parameter 'values' of type 'Span<string?>' in 'void HttpHeaders.ReadStoreValues<object?>(Span<string?> values, object? storeValue, HttpHeaderParser? parser, ref int currentIndex)' due to differences in the nullability of reference types. [/vmr/src/runtime/src/libraries/System.Net.Http/src/System.Net.Http.csproj::TargetFramework=net10.0-linux]
    /vmr/src/runtime/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs(1209,42): warning CS8620: Argument of type 'string[]' cannot be used for parameter 'values' of type 'Span<string?>' in 'void HttpHeaders.ReadStoreValues<string?>(Span<string?> values, object? storeValue, HttpHeaderParser? parser, ref int currentIndex)' due to differences in the nullability of reference types. [/vmr/src/runtime/src/libraries/System.Net.Http/src/System.Net.Http.csproj::TargetFramework=net10.0-linux]
    /vmr/src/runtime/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorExtensions.cs(2661,56): error CS0023: Operator '.' cannot be applied to operand of type 'void' [/vmr/src/runtime/src/libraries/System.Numerics.Tensors/src/System.Numerics.Tensors.csproj::TargetFramework=net10.0]
        4 Warning(s)
        1 Error(s)
```

I'm not sure if those are related.

jkotas commented:
> While we wait for @jkoritzinsky's fix to flow, let's look at the source-build failures:

These look related to ["first class Span" Roslyn work](https://github.com/dotnet/runtime/pull/108789).  cc @jjonescz

am11 commented:
If this is the only error, it can be avoided: https://github.com/dotnet/runtime/compare/main...am11:patch-19

jjonescz commented:
> These look related to ["first class Span" Roslyn work](https://github.com/dotnet/runtime/pull/108789).

Correct. Although I don't really follow how updating net9 to net10 causes vmr to start failing on the new compiler feature.

> If this is the only error

I would expect there to be more errors in the runtime repo when they consume the feature. This PR should contain mitigations for all of them: https://github.com/dotnet/runtime/pull/108789

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

ViktorHofer committed:
Update and rename 0001-Forward-NetCoreAppToolCurrentVersion-property-and-fi.patch to 0001-Update-hardcoded-linker-TFM.patch

ViktorHofer committed:
Update patches

ViktorHofer commented:
@am11 we will know more in an hour. This build now contains the runtime fix from the COM source generator. We can for sure apply your patch then. Thanks for helping out here.

ViktorHofer committed:
Update patch again...

ViktorHofer committed:
Add another runtime patch

ViktorHofer committed:
damn... update patches again

ViktorHofer committed:
and again...

ViktorHofer commented:
```
    /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS1620: Argument 2 must be passed with the 'out' keyword [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
    /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS0269: Use of unassigned out parameter 'task' [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
    /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,137): error CS0177: The out parameter 'task' must be assigned to before control leaves the current method [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
```

@jkoritzinsky

ViktorHofer commented:
The source-build leg is now finally passing but then errors out because of new prebuilts:

```
/vmr/repo-projects/Directory.Build.targets(829,5): warning : 4 new packages used not in baseline! See report at /vmr/artifacts/prebuilt-report/baseline-comparison.xml for more information. Package IDs are: [/vmr/repo-projects/dotnet.proj]
/vmr/repo-projects/Directory.Build.targets(829,5): warning : Microsoft.DotNet.ILCompiler.10.0.0-alpha.1.24528.7 [/vmr/repo-projects/dotnet.proj]
/vmr/repo-projects/Directory.Build.targets(829,5): warning : Microsoft.NET.ILLink.Tasks.10.0.0-alpha.1.24528.7 [/vmr/repo-projects/dotnet.proj]
/vmr/repo-projects/Directory.Build.targets(829,5): warning : Microsoft.NETCore.App.Runtime.linux-x64.10.0.0-alpha.1.24528.7 [/vmr/repo-projects/dotnet.proj]
/vmr/repo-projects/Directory.Build.targets(829,5): warning : runtime.placeholder-rid.Microsoft.DotNet.ILCompiler.10.0.0-alpha.1.24528.7 [/vmr/repo-projects/dotnet.proj]
```

cc @dotnet/source-build

MichaelSimons commented:
@ellahathaway, I think we should add the prebuilts to the tarball to unblock and then follow-up.  Can you handle this?  TIA

jkotas commented:
> I don't really follow how updating net9 to net10 causes vmr to start failing on the new compiler feature.

@jjonescz The failing leg is source build. Source build is a special flavor of the vmr build. Source build uses live version of all components to build the world. It rebuilds the world multiple times to get there - first with "last known good" SDK to get the live SDK build, and second time to with the live build of SDK created by the first build. The new compiler feature is in the live .NET 10 SDK build and that' why it started failing.

jkoritzinsky commented:
> ```
> 
>     /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS1620: Argument 2 must be passed with the 'out' keyword [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
> 
>     /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,250): error CS0269: Use of unassigned out parameter 'task' [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
> 
>     /vmr/src/runtime/artifacts/obj/cdacreader/Release/net10.0/linux-x64/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Interop.ComInterfaceGenerator/Microsoft.Diagnostics.DataContractReader.Legacy.IXCLRDataProcess2.cs(1046,137): error CS0177: The out parameter 'task' must be assigned to before control leaves the current method [/vmr/src/runtime/src/native/managed/cdacreader/src/cdacreader.csproj]
> 
> ```
> 
> 
> 
We need to update the SDK in the global.json for the VMR to one that has the fix. The fix just flowing isn't enough.

ViktorHofer committed:
Update SDKs

ViktorHofer commented:
This is now blocked on sdk's official build failing. @jkoritzinsky is taking a look.

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

ViktorHofer committed:
Remove now unnecessary patch

ellahathaway committed:
Update to new prebuilts tarball version

am11 commented on global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "9.0.100-rc.2.24474.11",
+    "dotnet": "10.0.100-alpha.1.24530.16",
```
```suggestion
    "dotnet": "10.0.100-alpha.1.24551.9",
```

am11 commented on src/SourceBuild/content/global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "9.0.100-rc.2.24474.11"
+    "dotnet": "10.0.100-alpha.1.24530.16"
```
```suggestion
    "dotnet": "10.0.100-alpha.1.24551.9"
```

jkoritzinsky committed:
Apply suggestions from code review

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

am11 commented:
Ah, VMR legs are building libs.tests, so we will need to resolve conflicts and merge https://github.com/dotnet/runtime/pull/108789 as well.

am11 commented:
(I'm working on it)

MichaelSimons commented:
@ellahathaway - I see some version updates were pushed which changed some prebuilt versions.  Can you update the baseline again?  TIA

ellahathaway committed:
Update prebuilts tarball

am11 commented:
@ViktorHofer could you please merge main to this PR branch as well without updating SDK? updating SDK version in global.json seems to invalidate prebuild (https://github.com/dotnet/source-build/issues/4715) and we really don't need it for this round.

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

am11 commented:
Good news: runtime build is clear!

Not so good news:
1. dotnet-watch tests have hardcoded TFM. We can switch it to use regex match: https://github.com/dotnet/sdk/pull/44604 (confirmed locally that it builds, testing on CI 😅)
2. aspnetcore needs a similar roslyn patch (will work on the patch next, should be similar to runtime's patch)

am11 commented:
Good news: runtime build is clear!

Not so good news:
1. dotnet-watch tests have hardcoded TFM. We can switch it to use regex match: https://github.com/dotnet/sdk/pull/44604 (testing on CI 😅)
2. aspnetcore needs a similar roslyn patch (will work on the patch next, should be similar to runtime's patch)

ViktorHofer commented:
@am11 given that the aspnetcore PR is already merged, let's just wait for the normal dependency flow and then update this PR again with latest main. aspnetcore is configured to flow every commit into sdk so after their official build completed, we will see the dependency update.

am11 commented:
> @am11 given that the aspnetcore PR is already merged, let's just wait for the normal dependency flow and then update this PR again with latest main. aspnetcore is configured to flow every commit into sdk so after their official build completed, we will see the dependency update.

@ViktorHofer thanks. Just a note that I wasn't able to update roslyn in aspnetcore, so the changes were done using search and replace (https://github.com/dotnet/aspnetcore/pull/58769#issuecomment-2453115466). If we find more issues in VMR, lets patch them here first and once we have the complete patch, we can then upstream it. Looks like we were not expecting these errors (shouldn't in use-code): https://github.com/dotnet/runtime/pull/108789#discussion_r1796995408. cc @jjonescz

am11 commented:
> @am11 given that the aspnetcore PR is already merged, let's just wait for the normal dependency flow and then update this PR again with latest main. aspnetcore is configured to flow every commit into sdk so after their official build completed, we will see the dependency update.

@ViktorHofer thanks. Just a note that I wasn't able to update roslyn in aspnetcore, so the changes were done using search and replace (https://github.com/dotnet/aspnetcore/pull/58769#issuecomment-2453115466). If we find more issues in VMR, lets patch them here first and once we have the complete patch, we can then upstream it. Looks like we were not expecting these errors (shouldn't fail in user code): https://github.com/dotnet/runtime/pull/108789#discussion_r1796995408. cc @jjonescz

ViktorHofer committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

am11 commented:
```
       C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : System.IO.FileNotFoundException: Could not load file or assembly 'System.Text.Json, Version=8.0.0.5, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : File name: 'System.Text.Json, Version=8.0.0.5, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error :    at Microsoft.NET.Sdk.WebAssembly.GenerateWasmBootJson.WriteBootJson(Stream output, String entryAssemblyName) [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error :    at Microsoft.NET.Sdk.WebAssembly.GenerateWasmBootJson.Execute() [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error :  [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : WRN: Assembly binding logging is turned OFF. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : Note: There is some performance penalty associated with assembly bind failure logging. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog]. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error :  [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]
        Done Building Project "C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj" (default targets) -- FAILED.
        Done Building Project "C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Server\BlazorWasmHosted60.Server.csproj" (Publish target(s)) -- FAILED.
```

Do we need https://github.com/dotnet/sdk/pull/44029/files?

Second failure is in `UnifiedBuild.Tests`:
```
A total of 1 test files matched the specified pattern.
Testhost process for source(s) '/vmr/artifacts/bin/Microsoft.DotNet.UnifiedBuild.Tests/Release/Microsoft.DotNet.UnifiedBuild.Tests.dll' exited with error: You must install or update .NET to run this application.
App: /vmr/artifacts/bin/Microsoft.DotNet.UnifiedBuild.Tests/Release/testhost.dll
Architecture: x64
Framework: 'Microsoft.NETCore.App', version '9.0.0-rc.2.24473.5' (x64)
.NET location: /vmr/.dotnet/
The following frameworks were found:
  10.0.0-alpha.1.24531.8 at [/vmr/.dotnet/shared/Microsoft.NETCore.App]
```

ViktorHofer committed:
Set rollforward policy for VMR test projects

am11 commented:
> ` C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\BlazorLegacy60\microsoft.net.sdk.webassembly.pack\10.0.0-alpha.1.24531.8\build\Microsoft.NET.Sdk.WebAssembly.Browser.targets(366,5): error : System.IO.FileNotFoundException: Could not load file or assembly 'System.Text.Json, Version=8.0.0.5, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified. [C:\h\w\AEEF095C\t\dotnetSdkTests\wiq1g4ih.kft\Publish60Host---0200F604\Client\BlazorWasmHosted60.Client.csproj]`

This one still persists. Assuming we still cannot update STJ to 8.0.5 (although msbuild has been updated which we were supposed to keep in sync with) maybe disable the blazor tests against an issue and keep going?

ViktorHofer commented:
I tried to fix the unified-build wpf repo failures with https://github.com/dotnet/wpf/pull/10028 but that doesn't seem to work.

ViktorHofer commented:
> This one still persists. Assuming we still cannot update STJ to 8.0.5 (although msbuild has been updated which we were supposed to keep in sync with) maybe disable the blazor tests against an issue and keep going?

Our CI images still use a VS installation without STJ/8.0.0.5.

am11 commented:
> Our CI images still use a VS installation without STJ/8.0.0.5.

Understood. Can we disable blazor and few other tests blocking the updates, add them to this existing tracking issue https://github.com/dotnet/sdk/issues/42920? There seems to be follow-up after this PR as-is, so hopefully it won't make a huge difference. 😅

ViktorHofer commented:
Sure. Any idea why this is only showing up in this PR?

am11 commented:
I think this is because now we are building with 10.0 (vs. 9.0-rc2 in main), and https://github.com/dotnet/runtime/commit/7318855838b7cf355a7491aadf4b773dd5914617 and similar changes are included only in 10.0.

ViktorHofer committed:
Disable Publish60Hosted_Works until 8.0.5 is available on .NET Framework

marcpopMSFT commented:
We haven't been given an ETA for when test images will include VS versions that have 8.0.5. Limiting the test to 17.12.0 is the right change.

marcpopMSFT commented:
@javiercn one of the blazor tests needs an update it looks like. Is that you who should take a look?
`Expected manifest Identity of ${ProjectPath}\Client\obj\Debug${Tfm}\blazor.publish.boot.json but found ${ProjectPath}\Client\obj\Debug${Tfm}\publish.blazor.boot.json.
Expected manifest OriginalItemSpec of ${ProjectPath}\Client\obj\Debug${Tfm}\blazor.publish.boot.json but found ${ProjectPath}\Client\obj\Debug${Tfm}\publish.blazor.boot.json.`

javiercn commented:
@marcpopMSFT I can take a look at it tomorrow.

mthalman committed:
Update prebuilts tarball

ViktorHofer commented on test/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/BlazorLegacyIntegrationTest60.cs:
```diff
@@ -44,7 +44,7 @@ public void Build60Hosted_Works()
         }
 
-        [PlatformSpecificFact(TestPlatforms.Windows)]
+        [RequiresMSBuildVersionFact("17.12", Reason = "Needs System.Text.Json 8.0.5")]
```
This change didn't work as the test now runs on non Windows platforms:

> Test Microsoft.NET.Sdk.BlazorWebAssembly.Tests.BlazorLegacyIntegrationTest60.Publish60Hosted_Works has failed.

@marcpopMSFT how do I use both the `RequiresMSBuildVersionFact` and `PlatformSpecificFact` logic in this repo? I would have assumed that the `SkipOnPlatform` attribute below makes sure that this doesn't run on linux or osx but apparently it does.

am11 replied:
Going by https://github.com/dotnet/sdk/blob/93788ad6cc8841e60e60a3ea780ed0a14f85ac48/test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPublishAnAotApp.cs#L438-L442 we can wrap the body in condition?

marcpopMSFT replied:
am11 is correct, we don't have a single Fact/Theory that limits both the msbuild version and the platform so you have to use an alternative check. So this test wasn't meant to be running then and that's the fix.

am11 replied:
Ah there is a better one:
```suggestion
        [WindowsOnlyRequiresMSBuildVersionFact("17.12", Reason = "Needs System.Text.Json 8.0.5")]
```

marcpopMSFT replied:
forgot we had that one, good catch.

marcpopMSFT committed:
switch to the windows only msbuild fact

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

lewing commented:
> @javiercn one of the blazor tests needs an update it looks like. Is that you who should take a look? `Expected manifest Identity of ${ProjectPath}\Client\obj\Debug${Tfm}\blazor.publish.boot.json but found ${ProjectPath}\Client\obj\Debug${Tfm}\publish.blazor.boot.json. Expected manifest OriginalItemSpec of ${ProjectPath}\Client\obj\Debug${Tfm}\blazor.publish.boot.json but found ${ProjectPath}\Client\obj\Debug${Tfm}\publish.blazor.boot.json.`

This is expected, we should just change the test here. cc @maraf

am11 commented:
Looks like it needs merge-from-main.

nagilson committed:
Merge branch 'main' into dev/Jason/Update-the-sdk-target-to-net10.0

am11 commented on test/Microsoft.NET.Sdk.BlazorWebAssembly.Tests/BlazorLegacyIntegrationTest60.cs:
```diff
@@ -44,7 +44,7 @@ public void Build60Hosted_Works()
         }
 
-        [PlatformSpecificFact(TestPlatforms.Windows)]
+        [WindowsOnlyRequiresMSBuildVersionFact("17.12", Reason = "Needs System.Text.Json 8.0.5")]
```
```suggestion
        [WindowsOnlyRequiresMSBuildVersionFact("17.12")] // Needs System.Text.Json 8.0.5
```

Commit:
Add Reason property to and update baselines

ViktorHofer committed:
Add wpf patch

ViktorHofer commented:
The added patch should fix the wpf SDK mismatch issue but I wouldn't be surprised if that repo needs an additional change due to BCL changes. Let's see.

ViktorHofer commented:
> D:\a\_work\1\vmr\.dotnet\sdk\10.0.100-alpha.1.24551.9\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1060: Error reading assets file: Error loading lock file 'D:\a\_work\1\vmr\src\wpf\artifacts\obj\Microsoft.NET.Sdk.WindowsDesktop.ArchNeutral\project.assets.json' : Could not load file or assembly 'System.Text.Json, Version=8.0.0.4, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified. [D:\a\_work\1\vmr\src\wpf\packaging\Microsoft.NET.Sdk.WindowsDesktop\Microsoft.NET.Sdk.WindowsDesktop.ArchNeutral.csproj]

Hmm that one is annoying. The wpf repo builds with desktop msbuild. The VS version that is part of the build image is `Preview 17.11.0-pre.2.1`:

![image](https://github.com/user-attachments/assets/08a24cf1-d540-42af-aba1-e14ac4e356aa)

That VS installation seems to be missing https://github.com/dotnet/msbuild/commit/7500f064233494f673100e3c1d260006792cf9c6 which updated the STJ binding redirects to 8.0.0.4. Apparently NuGet assemblies depend on the 8.0.0.4 version though. @rainersigwald ideas?

Btw I'm not sure why the windows.vs2022.preview image is lagging behind so much. It would be great if it could get updated to a VS17.12 version.

ViktorHofer commented:
Yes but the workaround is already checked into the Arcade.Sdk and runs by default. The binlog shows that the target runs but the ResolvePackageAssets task is still failing:

![image](https://github.com/user-attachments/assets/128895b7-b5a4-4ec9-9f07-4553e95652e2)

ViktorHofer committed:
Add Arcade patch

ViktorHofer committed:
Damn, I hate patches

ViktorHofer commented:
Now windowsdesktop (sigh). Still, all those failures are really helpful in making our infrastructure more robust.

ViktorHofer committed:
Add windowsdesktop patch

ViktorHofer commented:
A bit thanks to our community contributors @am11 and @kasperk81 for helping getting this in. You are amazing! 🎉

am11 commented:
Haha, classic move! Sometimes all it takes is a well-timed game-changing commit like `I hate patches.patch` to solve the mystery! 😜 Huge props to @ViktorHofer and others for the detective work! 🕵️‍♂️🏆
